base/keylist/keylist.go

```diff
@@ -0,0 +1,218 @@
+This is a different implementation of the [ordmap] package,
+that has separate slices for Values and Keys, instead of
+using a tuple list of both. The awkwardness of value access
+through the tuple is the major problem with ordmap.
+*/
+package keylist
+
+import (
+	"fmt"
+	"slices"
+)
+
+// TODO: probably want to consolidate ordmap and keylist
+
+// List implements an ordered list (slice) of Values,
+// with a map from a key (e.g., names) to indexes,
+// to support fast lookup by name.
+type List[K comparable, V any] struct { //types:add
+	// List is the ordered slice of items.
+	Values []V
+
+	Keys []K
+
+	// indexes is the key-to-index mapping.
+	indexes map[K]int
+}
+
+// New returns a new Key List.  Zero value
+// is usable without initialization, so this is
+// just a simple standard convenience method.
+func New[K comparable, V any]() *List[K, V] {
+	return &List[K, V]{}
+}
+
+func (kl *List[K, V]) newIndexes() {
+	kl.indexes = make(map[K]int)
+}
+
+// initIndexes ensures that the index map exists.
+func (kl *List[K, V]) initIndexes() {
+	if kl.indexes == nil {
+		kl.newIndexes()
+	}
+}
+
+// Reset resets the list, removing any existing elements.
+func (kl *List[K, V]) Reset() {
+	kl.Values = nil
+	kl.Keys = nil
+	kl.newIndexes()
+}
+
+// Add adds an item to the list with given key.
+// An error is returned if the key is already on the list.
+// See [AddReplace] for a method that automatically replaces.
+func (kl *List[K, V]) Add(key K, val V) error {
+	kl.initIndexes()
+	if _, ok := kl.indexes[key]; ok {
+		return fmt.Errorf("keylist.Add: key %v is already on the list", key)
+	}
+	kl.indexes[key] = len(kl.Values)
+	kl.Values = append(kl.Values, val)
+	kl.Keys = append(kl.Keys, key)
+	return nil
+}
+
+// AddReplace adds an item to the list with given key,
+// replacing any existing item with the same key.
+func (kl *List[K, V]) AddReplace(key K, val V) {
+	kl.initIndexes()
+	if idx, ok := kl.indexes[key]; ok {
+		kl.Values[idx] = val
+		kl.Keys[idx] = key
+		return
+	}
+	kl.indexes[key] = len(kl.Values)
+	kl.Values = append(kl.Values, val)
+	kl.Keys = append(kl.Keys, key)
+}
+
+// Insert inserts the given value with the given key at the given index.
+// This is relatively slow because it needs regenerate the keys list.
+// It returns an error if the key already exists because
+// the behavior is undefined in that situation.
+func (kl *List[K, V]) Insert(idx int, key K, val V) error {
+	if _, has := kl.indexes[key]; has {
+		return fmt.Errorf("keylist.Add: key %v is already on the list", key)
+	}
+
+	kl.Keys = slices.Insert(kl.Keys, idx, key)
+	kl.Values = slices.Insert(kl.Values, idx, val)
+	kl.newIndexes()
+	for i, k := range kl.Keys {
+		kl.indexes[k] = i
+	}
+	return nil
+}
+
+// ValueByKey returns the value corresponding to the given key,
+// with a zero value returned for a missing key. See [List.ValueByKeyTry]
+// for one that returns a bool for missing keys.
+func (kl *List[K, V]) ValueByKey(key K) V {
```

Comment:
Given that `ValueByKey` is one of the most common use cases, I feel like we should have a shorter name.

Reply:
`At`, `Value`, `ByKey` ?  I prefer `At`

Reply:
`At` done.